ehsm_kms_service/function.js

```diff
@@ -21,6 +31,158 @@ const base64_encode = (str) => new Buffer.from(str).toString('base64')
 const base64_decode = (base64_str) =>
   new Buffer.from(base64_str, 'base64').toString()
 
+/**
+ * @returns nonce_cache_timer
+ */
+const _nonce_cache_timer = () => {
+  const nonce_database = {}
+  const timer = setInterval(() => {
+    try {
+      for (const appid in nonce_database) {
+        let slice_index =
+          nonce_database[appid] &&
+          nonce_database[appid].findIndex((nonce_data) => {
+            return (
+              new Date().getTime() - nonce_data.nonce_timestamp >
+              NONCE_CACHE_TIME
+            )
+          })
+        if (slice_index > 0) {
+          nonce_database[appid] = nonce_database[appid].slice(slice_index)
+        }
+        if (slice_index == 0) {
+          delete nonce_database[appid]
+        }
+      }
+    } catch (error) {
+      res.send(_result(404, 'Not Fount', {}))
+      logger.error(JSON.stringify(error))
+    }
+  }, NONCE_CACHE_TIME / 2)
+  return { timer, nonce_database }
+}
+
+/**
+ * Clear expired cmks
+ * @returns _cmk_cache_timer
+ */
+const _cmk_cache_timer = (DB) => {
+  let nowTime = new Date().getTime()
+  let recent = new Date().setHours(CMK_LOOP_CLEAR_EXECUTION_TIME)
+  recent >= nowTime || (recent += 24 * 3600000)
+
+  const timer = setTimeout(() => {
+    setInterval(() => {
+      try {
+        const current_time = new Date().getTime() + CMK_EXPIRE_TIME_EXPAND
+        const query = {
+          selector: {
+            expireTime: { $lt: current_time },
+          },
+          fields: ['_id', '_rev'],
+          limit: 10000,
+        }
+        DB.partitionedFind('cmk', query)
+          .then((cmks_res) => {
+            console.log(cmks_res)
+            for (const cmk_item of cmks_res.docs) {
+              cmk_item._deleted = true
+            }
+            DB.bulk({ docs: cmks_res.docs }).catch((err) => {
+              console.log(err)
+            })
+          })
+          .catch((err) => {
+            console.log(err)
+          })
+      } catch (err) {
+        console.log(err)
+      }
+    }, CMK_LOOP_CLEAR_TIME)
+  }, recent - nowTime)
+
+  return { timer }
+}
+
+/**
+ *
+ * @param {object} napi_res
+ * @param {object} res
+ * @param {string} appid
+ * @param {object} DB
+ */
+function save_cmk(napi_res, res, appid, PAYLOAD, DB) {
+  try {
+    const creationDate = new Date().getTime()
+    const keyid = uuidv4().replace(/-/g, '')
+    let { keyspec, origin } = PAYLOAD
+
+    DB.insert({
+      _id: `cmk:${keyid}`,

```

Comment:
add list about the table format here.

Reply:
ok

Reply:
In **_id: `cmk:${keyid}`**,   Does "**cmk:**"  indicates a partition called cmk?

Reply:
yes